---
title: "changes"
author: "Yue Liu"
date: "2023-03-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

---
title: "R_assignment"
author: "Yue Liu"
date: "2023-03-11"
output: html_document
---


# set enviroment
```{r}
# install.packages("filesstrings")
library("tidyr")
library("tidyverse")
library("dplyr")
library("filesstrings")
```

# input data and do data inspections
```{r}
fang_et_al_genotypes <- read_tsv("./data/fang_et_al_genotypes.txt")
snp_position <- read_tsv("./data/snp_position.txt")
```

# data inspection

```{r}
typeof(fang_et_al_genotypes)
class(fang_et_al_genotypes)
str(fang_et_al_genotypes)

typeof(snp_position)
class(snp_position)
str(snp_position)
```


# select maize and teosinte SNP data based on groups
# assign slected rows into new variables
# remove the fisrt 3 columns selected rows 
```{r}
maize <- filter(fang_et_al_genotypes, Group %in% c('ZMMIL', 'ZMMLR', 'ZMMMR'))

teosinte <- filter(fang_et_al_genotypes, Group %in% c('ZMPBA', 'ZMPIL', 'ZMPJA'))
```




# prepare genotype data before merging with snp id

```{r}
maize <- select(maize, -(JG_OTU:Group))
maize_t <- t(maize) %>% as.data.frame(.) # transpose data 
maize_t <- tibble::rownames_to_column(maize_t, "Sample_ID") # convert the row names to first column
maize_t <- as_tibble(lapply(maize_t, gsub, pattern = "Sample_ID", replace = "SNP_ID")) 
columns <- maize_t[1,] # extract first row
colnames(maize_t) <- columns # use first row as column names
maize_t <- maize_t[-1,] # remove first row

teosinte <- select(teosinte, -(JG_OTU:Group))
teosinte_t <- t(teosinte) %>% as.data.frame(.)
teosinte_t <- tibble::rownames_to_column(teosinte_t, "Sample_ID")
teosinte_t <- as_tibble(lapply(teosinte_t, gsub, pattern = "Sample_ID", replace = "SNP_ID")) 
columns <- teosinte_t[1,] # extract first row
colnames(teosinte_t) <- columns # use first row as column names
teosinte_t <- teosinte_t[-1,] # remove first row
```


# extract SNP IDs from "snp_position.txt" and assign new variables
```{r}
snp_id <- select(snp_position, SNP_ID, Chromosome, Position) 
snp_id <- as.data.frame(snp_id)
```


# merge SNP data with SNP ID
```{r}
maize_merged <- inner_join(snp_id, maize_t, by = "SNP_ID")
teosinte_merged <- inner_join(snp_id, teosinte_t, by = "SNP_ID")
```



# replace missing dat by ? or -
```{r}
maize_merged_with_question <- as_tibble(lapply(maize_merged, gsub, pattern = "\\?", replacement = "\\?")) 
maize_merged_with_dash <- as_tibble(lapply(maize_merged, gsub, pattern = "\\?", replacement = "\\-"))
teosinte_merged_with_question <- as_tibble(lapply(teosinte_merged, gsub, pattern = "\\?", replacement = "\\?"))
teosinte_merged_with_dash <- as_tibble(lapply(teosinte_merged, gsub, pattern = "\\?", replacement = "\\-"))

maize_merged_with_question$Position <- as.double(maize_merged_with_question$Position) # change the type of Position column to numeric for further sorting
maize_merged_with_dash$Position <- as.double(maize_merged_with_dash$Position)
teosinte_merged_with_question$Position <- as.double(teosinte_merged_with_question$Position)
teosinte_merged_with_dash$Position <- as.double(teosinte_merged_with_dash$Position)

```


# creat directories for exporting maize and tosinte data
```{r}
dir.create('./maize_after_review')
dir.create('./teosinte_after_review')
```

# produce maize files grouped by chromosome with ascending position, missiing data is replaced by ?, save in ./maize_after_review

```{r}
for (i in 1:10) {subdata <- filter(maize_merged_with_question, Chromosome == i) %>% arrange(Position);
  filename <- paste("./maize_after_review/maize_ascending_chr", i, ".txt", sep = "");
   write.table(subdata, file = filename, quote = F, sep = "\t", row.names = F )
}
```


```{r}
maize_ascending_chr5 <- read.table("./maize_after_review/maize_ascending_chr5.txt") # view one of the results to double check
maize_ascending_chr5 <- read_tsv("./maize_after_review/maize_ascending_chr5.txt")
```



# produce maize files grouped by chromosome with descending position, missiing data is replaced by -, and save in ./maize_after_review

```{r}
for (i in 1:10) {subdata <- filter(maize_merged_with_dash, Chromosome == i) %>% arrange(desc(Position));
  filename <- paste("./maize_after_review/maize_descending_chr", i, ".txt", sep = "");
   write.table(subdata, file = filename, quote = F, sep = "\t", row.names = F )
}
```

```{r}
maize_descending_chr7 <- read_tsv("./maize_after_review/maize_descending_chr7.txt") # view one of the results to double check
```


# produce teosinte data grouped by chromose with ascending position, missing dat replaced by ?, save in ./teosinte_after_review

```{r}
for (i in 1:10) {subdata <- filter(teosinte_merged_with_question, Chromosome == i) %>% arrange(Position);
  filename <- paste("./teosinte_after_review/teosinte_ascending_chr", i, ".txt", sep = "");
  write.table(subdata, file = filename, quote = F, sep = "\t", row.names = F)
  
}
```


# produce teosinte files grouped by chromosome with descending position, missiing data is replaced by -, save in ./teosinte_after_review

```{r}
for (i in 1:10) {subdata <- filter(teosinte_merged_with_dash, Chromosome == i) %>% arrange(desc(Position));
  filename <- paste("./teosinte_after_review/teosinte_descending_chr", i, ".txt", sep = "");
  write.table(subdata, file = filename, quote = F, sep = "\t", row.names = F)
  
}
```



# data visualization


# set up
```{r}
library("ggplot2")
```


# SNP distribution on chromosomes

```{r}
# as the SNP positions in both species are the same, I just used maize data to do the ploting

# data processing before plotting
as <- maize_merged
as <- as[as$Chromosome %in% c(1:10),]
as <- as[as$Position != "multiple",]
as <- as[!is.na(as$Position),]
as$Position <- as.numeric(as.factor(as$Position))

as <- as %>%  group_by(Chromosome) %>% mutate(position.bin = cut(Position, 10, labels = FALSE)) # bin positions based on chromosomes

ggplot(data = as, aes(x=position.bin, fill= Chromosome)) +
  geom_bar()+
  facet_wrap(~ as.double(Chromosome), nrow = 2,ncol = 5)+
  xlab("Postion")
```



# SNP distributions accross chromosomes

```{r}

# SNP distribution across chromosomes - two species
as1 <- maize_merged
as1$class <- "maize"
as1 <- select(as1, Chromosome, class)
as2 <- teosinte_merged
as2$class <- "teosinte" 
as2 <- select(as2, Chromosome, class)
as <- rbind(as1,as2)
as <- filter(as, Chromosome %in% c(1:10))

ggplot(data = as) +
  geom_bar(aes(x=factor(Chromosome, levels = c(1:10)), fill=class),position="dodge") + 
  xlab("Chromosome")


# SNP distribution across chromosomes - maize
as <- maize_merged
as <- as[as$Chromosome %in% c(1:10),]
as <- as[as$Position != "multiple",]
as <- as[!is.na(as$Position),]
ggplot() + 
  geom_point(data = as, aes(x=factor(Chromosome,levels=c(1:10)), y=as.numeric(Position)/1000000), color = 'red', shape = 95, size = 5, alpha = 0.2) + 
  labs(title = 'SNP Distributions accross Chromosomes - Maize', x = 'Chromosome', y = 'Chromosomal Position(Mb)')  + 
  theme_bw() 

# SNP distribution across chromosomes - teosinte
as <- teosinte_merged
as <- as[as$Chromosome %in% c(1:10),]
as <- as[as$Position != "multiple",]
as <- as[!is.na(as$Position),]
ggplot() + 
  geom_point(data = as, aes(x=factor(Chromosome,levels=c(1:10)), y=as.numeric(Position)/1000000), color = 'red', shape = 95, size = 5, alpha = 0.2) + 
  labs(title = 'SNP Distributions accross Chromosomes - Teosinte', x = 'Chromosome', y = 'Chromosomal Position(Mb)')  + 
  theme_bw() 

```




# Missing data and amount of heterozygosity

```{r}

maize_aa <- maize_merged 
maize_aa <- filter(maize_merged, Chromosome %in% c(1:10))



ggplot(maize_aa) +
  geom_bar(aes(x=Chromosome))

ggplot(maize_aa) +
  geom_bar(aes(x=Chromosome, fill = Chromosome))

ggplot(maize_aa) +
  geom_bar(aes(x=Chromosome, fill = Chromosome), position = "fill")

```




```{r}
# read data
heterozygosity_maize <- fang_et_al_genotypes
heterozygosity_maize <- filter(heterozygosity_maize, Group %in% c('ZMMIL', 'ZMMLR', 'ZMMMR')) 


for (i in c("A/A","T/T", "C/C", "G/G")) {
  heterozygosity_maize <- as_tibble(lapply(heterozygosity_maize, gsub, pattern = i, replacement = "homo") )
} 

for (i in c("A/T", "A/G","A/C","T/A","T/C", "T/G","C/G","C/T","C/A","G/C", "G/A","G/T")) {
  heterozygosity_maize <- as_tibble(lapply(heterozygosity_maize, gsub, pattern = i, replacement = "hetero") )
}

heterozygosity_maize <- as_tibble(lapply(heterozygosity_maize, gsub, pattern = "\\?\\/\\?", replacement = "miss") )

heterozygosity_maize <- heterozygosity_maize %>% 
  pivot_longer(cols = -c ("Sample_ID", "JG_OTU","Group"),
               names_to = "SNP_ID",
               values_to = "snp_data"
               )
ggplot(heterozygosity_maize) +
  geom_bar(aes(x=Group, fill = snp_data))

ggplot(heterozygosity_maize) +
  geom_bar(aes(x=Group, fill = snp_data), position = "fill")



```

```{r}
#count SNP based on types
heterozygosity <- fang_et_al_genotypes 
heterozygosity$missing_data <- rowSums(heterozygosity == "?/?")
heterozygosity$AA <- rowSums(heterozygosity == "A/A")
heterozygosity$TT <- rowSums(heterozygosity == "T/T")
heterozygosity$CC <- rowSums(heterozygosity == "C/C")
heterozygosity$GG <- rowSums(heterozygosity == "G/G")
heterozygosity <- mutate(heterozygosity, homo = AA + TT + CC + GG)
heterozygosity <- mutate(heterozygosity, hetero = 983 - homo - missing_data)
heterozygosity <- select(heterozygosity, Sample_ID, Group, homo, hetero, missing_data) %>% as.data.frame()
maize

#heterozygosity <- mutate(heterozygosity, homo = homo/983, hetero = hetero/983, missing_data = missing_data/983)

ggplot(heterozygosity) +
  geom_bar(aes(x=Group, fill = Group))

ggplot(heterozygosity) +
  geom_bar(aes(x=Group, fill = ae1.4))

ggplot(heterozygosity) +
  geom_bar(aes(x=Group, fill = Group), position = "fill")
```


```


# my own visualization

```{r}

```


