---
title: "R_assignment"
author: "Yue Liu"
date: "2023-03-11"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# set enviroment
```{r}
install.packages("filesstrings")
library("tidyr")
library("tidyverse")
library("dplyr")
library("filesstrings")
```

# input data and do data inspections
```{r}
snp_data <- read_tsv("./data/fang_et_al_genotypes.txt", col_names = TRUE)
snp_position <- read_tsv("./data/snp_position.txt", col_names = TRUE)
```

# data inspection

```{r}
typeof(snp_data)
str(snp_data)
```

# data inspection

```{r}
typeof(snp_position)
str(snp_position)
```


# select maize and teosinte SNP data based on groups
# assign slected rows into new variables
# remove the fisrt 3 columns selected rows 
```{r}
maize <- filter(snp_data, Group %in% c('ZMMIL', 'ZMMLR', 'ZMMMR'))
maize <- select(maize, -(Sample_ID:Group))
as.list(maize)
teosinte <- filter(snp_data, Group %in% c('ZMPBA', 'ZMPIL', 'ZMPJA'))
teosinte <- select(teosinte, -(Sample_ID:Group))
as.list(maize) 
```


# transpose maize and teosinte SNP data, assign new variables, and change data structure
```{r}
maize_t <- t(maize) 
maize_t <- as.data.frame(maize_t)
teosinte_t <- t(teosinte)
teosinte_t <- as.data.frame(teosinte_t)
```


# extract SNP IDs from "snp_position.txt" and assign new variables
```{r}
snp_id <- select(snp_position, SNP_ID, Chromosome, Position)
snp_id <- as.data.frame(snp_id)
```


# merge SNP data with SNP ID
```{r}
maize_merged <- mutate(snp_id, maize_t)
teosinte_merged <- mutate(snp_id, teosinte_t)

```



# replace missing dat by ? or -
```{r}
maize_merged_with_q <- as_tibble(lapply(maize_merged, gsub, pattern = "\\?\\/\\?", replacement = "\\?"))
maize_merged_with_d <- as_tibble(lapply(maize_merged, gsub, pattern = "\\?\\/\\?", replacement = "\\-"))
teosinte_merged_with_q <- as_tibble(lapply(teosinte_merged, gsub, pattern = "\\?\\/\\?", replacement = "\\?"))
teosinte_merged_with_d <- as_tibble(lapply(teosinte_merged, gsub, pattern = "\\?\\/\\?", replacement = "\\-"))
```

# creat directories and make it tidy
```{r}
dir.create('./maize')
dir.create('./teosinte')
```

# produce maize files grouped by chromosome with ascending position, missiing data is replaced by ?, save in ./maize

```{r}
f <- maize_merged_with_q
 
for (i in 1:10){subdata=f[f$Chromosome==i,];subdata=subdata[order(subdata$Position),];of=paste('./maize/maize_ascending_chr', i,sep='_');colnames(subdata)=colnames(f);write.table(subdata,file=of,quote=F,sep='\t',row.names=F)}
```


maize_ascending_chr1 <- filter(maize_merged_with_q, Chromosome ==1) %>% arrange(Position)
maize_ascending_chr2 <- filter(maize_merged_with_q, Chromosome ==2) %>% arrange(Position)
maize_ascending_chr3 <- filter(maize_merged_with_q, Chromosome ==3) %>% arrange(Position)
maize_ascending_chr4 <- filter(maize_merged_with_q, Chromosome ==4) %>% arrange(Position)
maize_ascending_chr5 <- filter(maize_merged_with_q, Chromosome ==5) %>% arrange(Position)
maize_ascending_chr6 <- filter(maize_merged_with_q, Chromosome ==6) %>% arrange(Position)
maize_ascending_chr7 <- filter(maize_merged_with_q, Chromosome ==7) %>% arrange(Position)
maize_ascending_chr8 <- filter(maize_merged_with_q, Chromosome ==8) %>% arrange(Position)
maize_ascending_chr9 <- filter(maize_merged_with_q, Chromosome ==9) %>% arrange(Position)
maize_ascending_chr10 <- filter(maize_merged_with_q, Chromosome ==10) %>% arrange(Position)


# produce maize files grouped by chromosome with descending position, missiing data is replaced by -, and save in ./maize

```{r}
f <- maize_merged_with_d
 
for (i in 1:10){subdata=f[f$Chromosome==i,];subdata=subdata[order(subdata$Position),decrease=T];of=paste('./maize/maize_descending_chr', i,sep='_');colnames(subdata)=colnames(f);write.table(subdata,file=of,quote=F,sep='\t',row.names=F)}
```

```{r}
maize_descending_chr1 <- filter(maize_merged_with_d, Chromosome ==1) %>% arrange(desc(Position))
maize_descending_chr2 <- filter(maize_merged_with_d, Chromosome ==2) %>% arrange(desc(Position))
maize_descending_chr3 <- filter(maize_merged_with_d, Chromosome ==3) %>% arrange(desc(Position))
maize_descending_chr4 <- filter(maize_merged_with_d, Chromosome ==4) %>% arrange(desc(Position))
maize_descending_chr5 <- filter(maize_merged_with_d, Chromosome ==5) %>% arrange(desc(Position))
maize_descending_chr6 <- filter(maize_merged_with_d, Chromosome ==6) %>% arrange(desc(Position))
maize_descending_chr7 <- filter(maize_merged_with_d, Chromosome ==7) %>% arrange(desc(Position))
maize_descending_chr8 <- filter(maize_merged_with_d, Chromosome ==8) %>% arrange(desc(Position))
maize_descending_chr9 <- filter(maize_merged_with_d, Chromosome ==9) %>% arrange(desc(Position))
maize_descending_chr10 <- filter(maize_merged_with_d, Chromosome ==10) %>% arrange(desc(Position))
```

# produce teosinte data grouped by chromose with ascending position, missing dat replaced by ?, save in ./teosinte

```{r}
f <- teosinte_merged_with_q
 
for (i in 1:10){subdata=f[f$Chromosome==i,];subdata=subdata[order(subdata$Position),];of=paste('./teosinte/teosinte_ascending_chr', i,sep='_');colnames(subdata)=colnames(f);write.table(subdata,file=of,quote=F,sep='\t',row.names=F)}
```


```{r}
teosinte_ascending_chr1 <- filter(teosinte_merged_with_q, Chromosome ==1) %>% arrange(Position)
teosinte_ascending_chr2 <- filter(teosinte_merged_with_q, Chromosome ==2) %>% arrange(Position)
teosinte_ascending_chr3 <- filter(teosinte_merged_with_q, Chromosome ==3) %>% arrange(Position)
teosinte_ascending_chr4 <- filter(teosinte_merged_with_q, Chromosome ==4) %>% arrange(Position)
teosinte_ascending_chr5 <- filter(teosinte_merged_with_q, Chromosome ==5) %>% arrange(Position)
teosinte_ascending_chr6 <- filter(teosinte_merged_with_q, Chromosome ==6) %>% arrange(Position)
teosinte_ascending_chr7 <- filter(teosinte_merged_with_q, Chromosome ==7) %>% arrange(Position)
teosinte_ascending_chr8 <- filter(teosinte_merged_with_q, Chromosome ==8) %>% arrange(Position)
teosinte_ascending_chr9 <- filter(teosinte_merged_with_q, Chromosome ==9) %>% arrange(Position)
teosinte_ascending_chr10 <- filter(teosinte_merged_with_q, Chromosome ==10) %>% arrange(Position)
```

# produce teosinte files grouped by chromosome with descending position, missiing data is replaced by -, save in ./teosinte


```{r}
f <- teosinte_merged_with_d
 
for (i in 1:10){subdata=f[f$Chromosome==i,];subdata=subdata[order(subdata$Position),decrease=T];of=paste('./teosinte/teosinte_descending_chr', i,sep='_');colnames(subdata)=colnames(f);write.table(subdata,file=of,quote=F,sep='\t',row.names=F)}
```

```{r}
teosinte_descending_chr1 <- filter(teosinte_merged_with_d, Chromosome ==1) %>% arrange(desc(Position)) 
teosinte_descending_chr2 <- filter(teosinte_merged_with_d, Chromosome ==2) %>% arrange(desc(Position))
teosinte_descending_chr3 <- filter(teosinte_merged_with_d, Chromosome ==3) %>% arrange(desc(Position))
teosinte_descending_chr4 <- filter(teosinte_merged_with_d, Chromosome ==4) %>% arrange(desc(Position))
teosinte_descending_chr5 <- filter(teosinte_merged_with_d, Chromosome ==5) %>% arrange(desc(Position))
teosinte_descending_chr6 <- filter(teosinte_merged_with_d, Chromosome ==6) %>% arrange(desc(Position))
teosinte_descending_chr7 <- filter(teosinte_merged_with_d, Chromosome ==7) %>% arrange(desc(Position))
teosinte_descending_chr8 <- filter(teosinte_merged_with_d, Chromosome ==8) %>% arrange(desc(Position))
teosinte_descending_chr9 <- filter(teosinte_merged_with_d, Chromosome ==9) %>% arrange(desc(Position))
teosinte_descending_chr10 <- filter(teosinte_merged_with_d, Chromosome ==10) %>% arrange(desc(Position))


```



# data visualization


# set up
```{r}
install.packages("ggplot")
library("ggplot")
```


# 
```{r}
ggplot(data = "maize_merge" ) + geom_point(mapping = aes(x=Position, y = ))
```



